---
title: "RNAseq with edgeR"
author: "Aurore Comte"
date: "`r Sys.time()`"
output:
  html_document:
    code_folding: hide
    highlight: tango
    number_sections: yes
    theme: cerulean
    toc: yes
    toc_float:
      collapsed: yes
      smooth_scroll: yes
  pdf_document:
    toc: yes
editor_options:
  chunk_output_type: console
---

```{r knitrGlobalOpt}
knitr::opts_chunk$set(dev='CairoPNG')
options(knitr.table.format = "markdown")
```

# Load Packages

```{r include=FALSE}
.Last <- function() {
  cat("Now saving current session to output dir:", outDir, "\n")
  save.image(file = file.path(outDir, "session.RData"), compress = TRUE)
  cat("bye bye...\n")
}

suppressMessages(library('knitr', warn.conflict = FALSE, quietly = TRUE))
suppressMessages(library('edgeR', warn.conflict = FALSE, quietly = TRUE))
suppressMessages(library('limma', warn.conflict = FALSE, quietly = TRUE))
suppressMessages(library('RColorBrewer', warn.conflict = FALSE, quietly = TRUE))
suppressMessages(library('mixOmics', warn.conflict = FALSE, quietly = TRUE))
suppressMessages(library('baySeq', warn.conflict = FALSE, quietly = TRUE))
suppressMessages(library('Cairo', warn.conflict = FALSE, quietly = TRUE))
suppressMessages(library('plotly', warn.conflict = FALSE, quietly = TRUE))
suppressMessages(library('Rsubread', warn.conflict = FALSE, quietly = TRUE))
```

```{r include=FALSE}
#http://www.bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf

#opts_knit$set(root.dir = snakemake@params[["out_dir"]])

##################### LOADING PARAMS FROM SNAKEMAKE OBJECT ######################
gcsv_hisat <- snakemake@input[["gcsv_hisat"]]
sample_info <- snakemake@input[["sample_info"]]
comparaisons_file <- snakemake@input[["de_comparisons_file"]]
gtf_file <- snakemake@input[["gtf_file"]]

outDir <- snakemake@params[["out_dir"]]
invisible(dir.exists(outDir) || dir.create(path = outDir, mode = "770"))

normCount_file <- snakemake@params[["normCount"]]
counts_file <- snakemake@params[["Counts"]]
resDE_file <- snakemake@params[["resDE"]]

cutoff_cpm <- snakemake@params[["cutoff_cpm"]]
cutoff_nb_echantillons <- snakemake@params[["cutoff_nb_echantillons"]]
thres_FDR <- snakemake@params[["thres_FDR"]]
thres_logFC <- snakemake@params[["thres_logFC"]]
#########################################################################
```
```{r include=FALSE}
#test

#gcsv_hisat <- "/home/comtea/Documents/PHIM/BURKADAPT/RNAja/output/3_count/STRINGTIE/HISAT_gene_count_matrix.csv"
#sample_info <- "/home/comtea/Documents/PHIM/BURKADAPT/RNAja/DATA/sample_info_all.txt"
#comparaisons_file <- "/home/comtea/Documents/PHIM/BURKADAPT/RNAja/DATA/treatmentsComparisons_all.csv"
#outDir <- "/home/comtea/Documents/PHIM/BURKADAPT/RNAja/test"
#invisible(dir.exists(outDir) || dir.create(path = outDir, mode = "770"))
#normCount_file <- "/home/comtea/Documents/PHIM/BURKADAPT/RNAja/test/normCounts"
#counts_file <- "/home/comtea/Documents/PHIM/BURKADAPT/RNAja/test/Counts"
#resDE_file <- "/home/comtea/Documents/PHIM/BURKADAPT/RNAja/test/resde"
#gtf_file="/home/comtea/Documents/PHIM/BURKADAPT/RNAja/DATA/ref/msu7.gtf"

#cutoff_cpm <- 5
#cutoff_nb_echantillons <- 2
#thres_FDR <- 0.01
#thres_logFC <- 1.5
```

```{r include = FALSE}
# test en fonction des paires
comp = read.csv(comparaisons_file)
pair_txt = matrix(nrow=nrow(comp),ncol=2)
for (lines in 1:nrow(comp)){
  pair_txt[lines,1] = lines
  pair_txt[lines,2] = knit_expand(text= "\n\n## {{comp[lines,1]}},{{comp[lines,2]}}\n\n")
}
```

# creation DGEList object

```{r}
  pheno_data <- read.csv(sample_info)
  sampleInfo <- data.frame(fileName= pheno_data$SampleName, condition = pheno_data$Treatment, replicate= pheno_data$Experiment)
  rawCountTable <- as.matrix(read.csv(gcsv_hisat,row.names="gene_id"))

  dgList <- DGEList(rawCountTable, group=sampleInfo$condition, genes=rownames(rawCountTable))
  summary(cpm(dgList))

```
# filtering and normalization

Genes with very low counts across all libraries provide little evidence for differential expression.
In the biological point of view, a gene must be expressed at some minimal level before it is likely
to be translated into a protein or to be biologically important.
The filterByExpr function keeps rows that have worthwhile counts in a minumum number of samples

```{r}
keep <- filterByExpr(dgList)
table(keep)
dgList <- dgList[keep, , keep.lib.sizes=FALSE]
summary(cpm(dgList))
```
```{r}
dgList <- calcNormFactors(dgList, method="TMM")

eff.lib.size <- dgList$samples$lib.size*dgList$samples$norm.factors
normCounts <- cpm(dgList)

write.table(dgList$counts,file=counts_file,sep=";")
write.table(normCounts, file=normCount_file, sep=";")
```

# exploration and QC

```{r}
plotMDS(dgList)
```


# design matrix and estimating the dispersion

```{r}
Treat = factor(sampleInfo$condition)
Treat = relevel(Treat, ref="FC")
design <- model.matrix(~Treat)
logFC <- predFC(dgList,design,prior.count=1,dispersion=0.05)
cor(logFC[,4:6])

design <- model.matrix(~Time+Treat)
rownames(design) <- colnames(dgList)
design

library(statmod)

dgList <- estimateDisp(dgList, design, robust=TRUE)
fit <- glmQLFit(dgList, design, robust=TRUE)
plotQLDisp(fit)

qlf <- glmQLFTest(fit, coef=2:3)
topTags(qlf)

FDR <- p.adjust(qlf$table$PValue, method="BH")
sum(FDR < 0.05)
qlf <- glmQLFTest(fit)
topTags(qlf)
top <- rownames(topTags(qlf))
cpm(dgList)[top,]
summary(decideTests(qlf))
plotMD(qlf)
abline(h=c(-1,1), col="blue")
```


# DE and result



```{r, echo=FALSE,include = FALSE}
# You need this code to conduct the magic dependences attaching...
DT::datatable(matrix())
```

```{r, results='asis', echo=FALSE, message = FALSE, warning = FALSE}
for (lines in 1:nrow(comp)){
  cat(paste(knit(text = pair_txt[lines,2]), collapse = '\n'))

  dgeTest <- exactTest(dgList, pair=c(comp[lines,1],comp[lines,2]))
  dgeTest

  cat("\n\nplot an histogram of unadjusted p-values\n\n")

  hist(dgeTest$table[,"PValue"], breaks=50, main = "", xlab="pvalue")

  resNoFilt <- topTags(dgeTest, n=nrow(dgeTest$table))

  cat("\n\nnumber of differencialy expressed genes (risk: 1%)\n\n")

  sum(resNoFilt$table$FDR < thres_FDR)

  sigDownReg <- resNoFilt$table[resNoFilt$table$FDR<thres_FDR,]
  sigDownReg <- sigDownReg[order(sigDownReg$logFC),]
   cat(knitr::knit_print((DT::datatable(sigDownReg))))

  resDE_file_name = paste0(resDE_file,'_',comp[lines,1],"-",comp[lines,2])
  write.table(resNoFilt, file=resDE_file_name, sep=";")

  y <- cpm(dgList, log=TRUE, prior.count = 1)
  selY <- y[rownames(resNoFilt$table)[resNoFilt$table$FDR<thres_FDR & abs(resNoFilt$table$logFC)>thres_logFC],]
  cimColor <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)[255:1]
  finalHM <- cim(t(selY), color=cimColor, symkey=FALSE, margins=c(0.1,0.1))

  cat("\n\ncreate a MA plot with 1% differentially expressed genes\n\n")
  plotSmear(dgeTest, de.tags = rownames(resNoFilt$table)[which(resNoFilt$table$FDR < thres_FDR)], xlab = "Average LogCPM", ylab="logFC")

  cat("\n\n#### gene clustering \n\n")

  cat("We provide a cluster membership for each gene\n\n")

  plot(finalHM$ddc, leaflab="none")
  abline(h=10, lwd=2, col="pink")

  geneClust <- cutree(as.hclust(finalHM$ddc), h=10)

  x = matrix(nrow = length(unique(geneClust)), ncol = 2)
  for (i in 1:length(unique(geneClust))){
    x[i,1] = i
    x[i,2] = paste0(names(which(geneClust==i)), collapse = ", ")
  }
  cat(knitr::knit_print(DT::datatable(x)))
}
```